README.md
# HR_System
Python Certification 310 HR System Application


history.txt
9fd451c Added more tests
af92d7e Added more test cases to test harness
b73547b Added more test cases to test harness
f3620eb Added Unit Tests to Project
32d8883 Added Review Report and Export
cd71a7a Added employee edit functionality
aaba10f Added import CSV functionality and more Error handeling. Also cleaned up unused code
f5cf6fa refactored some functions
a6313d5 Creating report, displaying current employees and adding new employee added to project
1933a75 Added Save functionality but unable to get first column name
0701c0a Created Dictionary of lists to store data with auto incrimanting EmpId
767ef62 added menu option
666c398 Initial Commit
b81e3b7 Initial commit


Test_HR_System.py
# -*- coding: utf-8 -*-
#!/usr/bin/env python3
# ------------------------------------------#
# Title: Test_HR_System.py
# Desc: Assignment HR System - Create test cases for the HR System application
# Change Log: (Who, When, What)
# AHernandez, 2021-May-24, Created Unit Test File for HR System application
# AHernandez, 2021-May-29, Added more test cases to Test Harness
# ------------------------------------------#
import HR_System
from io import StringIO
#from io import StringIO
dictRow = {4: ['Andrew Smith', '123 W. Help', '123-45-6789', '1980-04-04', 'Struggling with this', '2010-04-04', '2100-01-01']}

def test_append_employee():
    expected = dictRow
    actual = HR_System.DataProcessor.append_employee('Andrew Smith', '123 W. Help', '123-45-6789', '1980-04-04', 'Struggling with this', '2010-04-04', '2100-01-01',4)
    assert expected == actual
    
def test_append_employee2():
    expected = dictRow
    actual = HR_System.DataProcessor.append_employee('Andrew Smith', '123 W. Help', '123-45-6789', '1980-04-04', 'Struggling with this', '2010-04-04', '2100-01-01',3)
    assert expected != actual
    
def test_find_emp_record():
    dictRow = {4: ['Andrew Smith', '123 W. Help', '123-45-6789', '1980-04-04', 'Struggling with this', '2010-04-04', '2100-01-01']}
    empOpt = 4
    expected = 4 
    actual = HR_System.DataProcessor.find_emp_record(empOpt,dictRow)
    assert expected == actual

def test_find_emp_record2(): 
    empOpt = 3
    expected = 4 
    actual = HR_System.DataProcessor.find_emp_record(empOpt,dictRow)
    assert expected != actual
    
def test_update_emp_record():
    dictRow = {3: ['Andrew Smith', '123 W. Help', '123-45-6789', '1980-04-04', 'Struggling with this', '2010-04-04', '2100-01-01']}
    expected = dictRow
    actual = HR_System.DataProcessor.update_emp_record('Andrew Smith', '123 W. Help', '123-45-6789', '1980-04-04', 'Struggling with this', '2010-04-04', '2100-01-01',3,dictRow)
    assert expected == actual
    
def test_read_file():
    lstTbl = []
    dictRow = {}
    expected = {'1': ['Andrew Hernandez', '131 N American St', '432-23-1234', '1980-06-07', 'Data Engineer', '2017-04-07', '2100-01-01'], '2': ['Juree', '234.SomeWhereElse', '848-12-4321', '1982-05-26', 'Sys Admin', '2019-03-01', '2020-03-01'], '3': ['Gia Hernandez', '123 W. Somewhere', '123-12-1234', '2013-09-23', 'Student', '2017-09-01', '2100-01-01']}
    actual = HR_System.FileProcessor.read_file('Employee.csv',dictRow,lstTbl)
    assert expected == actual

def test_export_to_text(monkeypatch):
    tblExport = StringIO('Yes')
    monkeypatch.setattr('sys.stdin',tblExport)
    assert HR_System.IO.export_to_text() == 'yes'  
    
def test_export_to_text2(monkeypatch):
    tblExport = StringIO('NO')
    monkeypatch.setattr('sys.stdin',tblExport)
    assert HR_System.IO.export_to_text() == 'no'     
    
def test_export_to_text3(monkeypatch):
    tblExport = StringIO('NO')
    monkeypatch.setattr('sys.stdin',tblExport)
    assert HR_System.IO.export_to_text() != 'Yes'      
    
def test_select_emp_edit(monkeypatch):
    edit_emp = StringIO('1')
    monkeypatch.setattr('sys.stdin',edit_emp)
    assert HR_System.IO.select_emp_edit() != '1'     

def test_select_emp_edit2(monkeypatch):
    edit_emp = StringIO('1')
    monkeypatch.setattr('sys.stdin',edit_emp)
    assert HR_System.IO.select_emp_edit() == 1  
      
def test_file_status(monkeypatch):
    file_status = StringIO('Yes')
    monkeypatch.setattr('sys.stdin',file_status)
    assert HR_System.Presentation.file_status() == 'yes'  
    
def test_file_status2(monkeypatch):
    file_status = StringIO('NO')
    monkeypatch.setattr('sys.stdin',file_status)
    assert HR_System.IO.export_to_text() == 'no'        
    
    
    
    
    
    
    
    
    
    
    
    

HR_System.py
# -*- coding: utf-8 -*-
#!/usr/bin/env python3
# ------------------------------------------#
# Title: HR_System.py
# Desc: Assignment HR System - Create the HR System application
# Change Log: (Who, When, What)
# AHernandez, 2021-May-24, Created File for HR System application and skeletal structure 
# AHernandez, 2021-May-25, Added data structure and input functions 
# AHernandez, 2021-May-25, Added entering new user, exporting to csv, and printing table
# AHernandez, 2021-May-26, added exporting table to text as a report  
# AHernandez, 2021-May-27, Added importing file functionality  
# AHernandez, 2021-May-28, Added Edit employee functionality
# AHernandez, 2021-May-29, Added Review Report and export
# ------------------------------------------#
import time
import re
from datetime import datetime 
import pandas as pd
import pathlib
import csv
from prettytable import PrettyTable

strChoice = '' 
lstTbl = []
dictRow = {}
menuOptions = '\nMenu:\n\t[R] Create Upcoming Review Report\n\t[L] Load Employee File\n\t[S] Save Current List\n\t[C] Current Employees \
                \n\t[P] Previous Employees\n\t[A] Add Employee\n\t[E] Edit Current Employee\n\t[X] Quit'
strMenuOption = 'r','l','s','c','p','a','e','x'
strFileName = 'Employee.csv'
strRegexSSN = '^(?!000|666)[0-8][0-9]{2}-(?!00)[0-9]{2}-(?!0000)[0-9]{4}$'
fieldnames = ['EmpId','Name', 'Address', 'SSN', 'DOB', 'jobTitle', 'startDate', 'EndDate']
ptable = ''

# -- PROCESSING -- #
class DataProcessor:
    def append_employee(Name, Address, SSN ,DOB, jobTitle, startDate,EndDate, empId):
        dictRow[empId] = [Name, Address, SSN ,DOB, jobTitle, startDate, EndDate]
        return dictRow
    
    def find_current_employee(dictRow):
        try:
            for key, val in dictRow.items():
                table = PrettyTable(fieldnames)
                for key, val in dictRow.items():
                    if datetime.now() < datetime.strptime(val[6], '%Y-%m-%d'):
                        table.add_row([key, *val])
            return table 
        except UnboundLocalError:
            print('The table is empty.  There is nothing to print.')   
            
    def find_previous_employee(dictRow):
        try:
            for key, val in dictRow.items():
                table = PrettyTable(fieldnames)
                for key, val in dictRow.items():
                    if datetime.now() > datetime.strptime(val[6], '%Y-%m-%d'):
                        table.add_row([key, *val])
            return table 
        except UnboundLocalError:
            print('The table is empty.  There is nothing to print.')
    
    def find_emp_record(empOpt,dictRow):
        try:
            for key in dictRow.keys():
                if empOpt == int(key):
                    empId = int(key)
            return empId
        except UnboundLocalError:
            print(f'Employee Id {empOpt} does not exist')
            
    def update_emp_record(Name, Address, SSN ,DOB, jobTitle, startDate,EndDate, empId,dictRow):
        for key in dictRow.keys():
            if int(key) == empId:
                dictRow[key] = [Name, Address, SSN ,DOB, jobTitle, startDate, EndDate]
        return dictRow
    
    def create_review_report(dicRow):
        try:
            for key, val in dictRow.items():
                table = PrettyTable(fieldnames)
                for key, val in dictRow.items():
                    if 0 <= (datetime.now().month - datetime.strptime(val[5], '%Y-%m-%d').month) <= 3 and datetime.now() < datetime.strptime(val[6], '%Y-%m-%d'):
                        table.add_row([key, *val])
            return table 
        except UnboundLocalError:
            print('The table is empty.  There is nothing to print.')    
            
# -- FILE PROCESSING -- #         
class FileProcessor:
    def read_file(file_name, dictRow,lstTbl):
        dictRow.clear()
        try:
            with open(file_name, mode='r') as file:
                next(file)
                for line in csv.reader(file):
                    tup = line[1],line[2],line[3],line[4],line[5],line[6],line[7]
                    lstTbl = list(tup)
                    dictRow[line[0]] = lstTbl
            print(f'Loading {file_name}: \n')
        except (FileNotFoundError,OSError):
            print(f'There is no file named {file_name}.')
        return dictRow
    
    def write_file(file_name,dictRow):         
        strOpt = Presentation.file_status()
        if strOpt == 'yes':
            try:
                (pd.DataFrame.from_dict(data=dictRow, orient='index',)
                 .to_csv(file_name, mode='w'))
            except ValueError:
                print('There is no Employee data')
            except PermissionError:
                print('The file is open, Please close and try again.')
        return None

    def export_current_table(tblex, table):
        if tblex == 'yes':
            with open('CurrentEmployeeTable.txt', 'w') as table_export:
                table_export.write(str(table))
    
    def export_previous_table(tblex, table):
        if tblex == 'yes':
            with open('PreviousEmployeeTable.txt', 'w') as table_export:
                table_export.write(str(table)) 
                
    def export_review_report(tblex, table):
        if tblex == 'yes':
            with open('UpcomingReviewReport.txt', 'w') as table_export:
                table_export.write(str(table))   

# -- INPUT/OUTPUT -- #
class IO:
    def choice(menuOptions):
        choice = input("Please select an Option: ").lower().strip()
        try:
            if choice.lower() in (menuOptions):
                return choice
            else:
                raise
        except RuntimeError:
            print("Ivalid option selected. Please try again\n")
            
    def add_new_employee(empId):
        
        empId = len(dictRow) + 1 
        strName = input('Enter an employees name: ').strip()  
        strAddress = input('Enter an employees Address: ').strip()
        
        while True:
            strSSN = input('Enter a SSN(xxx-xx-xxxx): ').strip() 
            try:
                if (re.search(strRegexSSN, strSSN)):
                    break
                else:
                    raise
            except: 
                print('Please enter a valid SSN in the following format xxx-xx-xxxx')
                
        while True:
            dateDOB = input('Enter a Date of Birth(YYYY-MM-DD): ').strip() 
            try:
                datetime.strptime(dateDOB, '%Y-%m-%d')
                break
            except ValueError:
                print("Incorrect DoB format, It should be YYYY-MM-DD")
                
        strJobTitle = input('Enter a Job Title: ').strip()
        
        while True:
            dateStartDate = input('Enter a Start Date(YYYY-MM-DD): ').strip() 
            try:
                datetime.strptime(dateStartDate, '%Y-%m-%d')
                break
            except ValueError:
                print("Incorrect Start Date format, It should be YYYY-MM-DD") 
        dateEndDate = '2100-01-01'               
        return strName, strAddress, strSSN ,dateDOB, strJobTitle, dateStartDate, dateEndDate, empId
    
    def export_to_text():
        tblExport = input('Would you like to export this table? (Yes/No) ').strip().lower()
        return tblExport
    
    def select_emp_edit():
        try:
            edit_emp = int(input('Select an EmployeeId to edit :'))
            return edit_emp
        except ValueError:
            print('Please enter an 10 base value')
            
    def edit_curent_employee():
                    
        strName = input('Update employee name: ').strip()  
        strAddress = input('Update employee Address: ').strip()
        
        while True:
            strSSN = input('Update SSN(xxx-xx-xxxx): ').strip() 
            try:
                if (re.search(strRegexSSN, strSSN)):
                    break
                else:
                    raise
            except: 
                print('Please enter a valid SSN in the following format xxx-xx-xxxx')
                
        while True:
            dateDOB = input('Update Date of Birth(YYYY-MM-DD): ').strip() 
            try:
                datetime.strptime(dateDOB, '%Y-%m-%d')
                break
            except ValueError:
                print("Incorrect DoB format, It should be YYYY-MM-DD")
                
        strJobTitle = input('Update Job Title: ').strip()
        
        while True:
            dateStartDate = input('Update Start Date(YYYY-MM-DD): ').strip() 
            try:
                datetime.strptime(dateStartDate, '%Y-%m-%d')
                break
            except ValueError:
                print("Incorrect Start Date format, It should be YYYY-MM-DD") 
        while True:
            dateEndDate = input('Update End Date(YYYY-MM-DD): ').strip() 
            try:
                datetime.strptime(dateEndDate, '%Y-%m-%d')
                break
            except ValueError:
                print("Incorrect End Date format, It should be YYYY-MM-DD")                  
        return strName, strAddress, strSSN ,dateDOB, strJobTitle, dateStartDate, dateEndDate       
            
# -- PRESENTATION -- #
class Presentation:    
    def file_status():
        print('You are about to write new data to the file.')
        strOpt = input('Would you like to proceed? (Yes/No) ').strip().lower()
        return strOpt
    
    def file_state(strFileName):
        print('The File {} does not exists!'.format(strFileName))
        print('\n') 
    
    def show_current_employee(table):
        if table != None:
            print (table)                
    
    def show_previous_employee(table):
        if table != None:
            print (table)                

    def review_message():
        print('\nCurrent employee(s) that are due for a review.  Be Generous!!!')
# -- Main -- #
class Main:    
    def main():
        global intEmpId
        intEmpId = 1
        file =  pathlib.Path(strFileName)
        if file.exists():
            FileProcessor.read_file(strFileName, dictRow,lstTbl)
        else:
            Presentation.file_state(strFileName)
        try:
            while True:
                print(menuOptions)
                strChoice = IO.choice(strMenuOption)
                if strChoice == 'x':
                    break
                elif strChoice == 'a':
                    strName, strAddress, strSSN ,dateDOB, strJobTitle, dateStartDate,dateEndDate, intEmpId = IO.add_new_employee(intEmpId)
                    DataProcessor.append_employee(strName, strAddress, strSSN 
                                                  ,dateDOB, strJobTitle, dateStartDate,dateEndDate, intEmpId)
                elif strChoice == 's':
                    FileProcessor.write_file(strFileName, dictRow)
                elif strChoice == 'l':
                    FileProcessor.read_file(strFileName, dictRow, lstTbl)
                elif strChoice == 'c':
                    ptable = DataProcessor.find_current_employee(dictRow)
                    Presentation.show_current_employee(ptable)
                    if ptable != None:
                        tblExport = IO.export_to_text()
                        FileProcessor.export_current_table(tblExport,ptable)
                elif strChoice == 'p':
                    ptable = DataProcessor.find_previous_employee(dictRow)
                    Presentation.show_previous_employee(ptable)
                    if ptable != None:
                        tblExport = IO.export_to_text()
                        FileProcessor.export_previous_table(tblExport,ptable)
                elif strChoice == 'e':
                    ptable = DataProcessor.find_current_employee(dictRow)
                    Presentation.show_current_employee(ptable)
                    edit_emp = IO.select_emp_edit()
                    intEmpId = DataProcessor.find_emp_record(edit_emp, dictRow)
                    if intEmpId != None:
                        strName, strAddress, strSSN ,dateDOB, strJobTitle, dateStartDate,dateEndDate = IO.edit_curent_employee()
                        DataProcessor.update_emp_record(strName, strAddress, strSSN ,dateDOB, strJobTitle, dateStartDate,dateEndDate,intEmpId,dictRow) 
                elif strChoice == 'r':
                    ptable = DataProcessor.create_review_report(dictRow)
                    Presentation.review_message()
                    Presentation.show_current_employee(ptable)
                    if ptable != None:
                        tblExport = IO.export_to_text()
                        FileProcessor.export_review_report(tblExport,ptable)
                    
        except KeyboardInterrupt:
            print('\nThe user has caused a keyboard interruption\nThe program will now close!')
            time.sleep(2)

if __name__ == '__main__':
    Main.main()
